pywr/parameters/control_curves.py

```diff
@@ -3,79 +3,295 @@
-        self.above_curve_upper_bounds = above_curve_upper_bounds
-        self.above_curve_lower_bounds = above_curve_lower_bounds
-        self.below_curve_upper_bounds = below_curve_upper_bounds
-        self.below_curve_lower_bounds = below_curve_lower_bounds
-        self._update_variable_properties()
-
-    def _update_variable_properties(self):
-        size = 0
-        lower_bounds = []
-        upper_bounds = []
-        if self.above_curve_upper_bounds is not None and self.above_curve_lower_bounds is not None:
-            size += 1
-            lower_bounds.append(self.above_curve_lower_bounds)
-            upper_bounds.append(self.above_curve_upper_bounds)
-
-        if self.below_curve_upper_bounds is not None and self.below_curve_lower_bounds is not None:
-            size += 1
-            lower_bounds.append(self.below_curve_lower_bounds)
-            upper_bounds.append(self.below_curve_upper_bounds)
-
-        self.size = size
-        self._lower_bounds = np.array(lower_bounds)
-        self._upper_bounds = np.array(upper_bounds)
+        if upper_bounds is not None:
+            if self.values is None:
+                raise ValueError('Upper bounds can only be specified if `values` is not `None`.')
+            if len(upper_bounds) != nvalues:
+                raise ValueError('Length of upper_bounds should be one more than the number of '
+                                 'control curves ({}).'.format(nvalues))
+            self._upper_bounds = np.array(upper_bounds)
+
+        if lower_bounds is not None:
+            if self.values is None:
+                raise ValueError('Lower bounds can only be specified if `values` is not `None`.')
+            if len(lower_bounds) != nvalues:
+                raise ValueError('Length of lower_bounds should be one more than the number of '
+                                 'control curves ({}).'.format(nvalues))
+            self._lower_bounds = np.array(lower_bounds)
+
+    @classmethod
+    def load(cls, model, data):
+        control_curves = super(ControlCurveParameter, cls)._load_control_curves(model, data)
+        storage_node = super(ControlCurveParameter, cls)._load_storage_node(model, data)
+
+        parameters = None
+        values = None
+        if 'values' in data:
+            values = load_parameter_values(model, data)
+        elif 'parameters' in data:
+            # Load parameters
+            parameters_data = data['parameters']
+            parameters = []
+            for pdata in parameters_data:
+                parameters.append(load_parameter(model, pdata))
+
+        return cls(control_curves, values=values, parameters=parameters, storage_node=storage_node)
 
     def value(self, ts, scenario_index):
         i = scenario_index.global_id
-        control_curve = self.control_curve.value(ts, scenario_index)
         node = self.node if self.storage_node is None else self.storage_node
```

Comment:
`self.node` is the `Node` the parameter is attached to, which could be different  to `storage_node` so I don't think that's a good idea. It could be set to a different private attribute e.g. `self._storage_node`